File: packages/core-ts/src/reliance/catalog.ts
// SPDX-License-Identifier: Apache-2.0
import { createHash } from 'node:crypto';
import type { DocumentLoader } from '../types.js';

export type CatalogErrorCode =
  | 'DUPLICATE_RESOURCE'
  | 'INTEGRITY_MISMATCH'
  | 'INVALID_RESOURCE'
  | 'RESOURCE_NOT_FOUND'
  | 'RESOURCE_BUDGET_EXCEEDED';

export class CatalogError extends Error {
  constructor(public readonly code: CatalogErrorCode, message: string) {
    super(message);
    this.name = 'CatalogError';
  }
}

export interface StaticResourceInput {
  uri: string;
  mediaType: string;
  bytes: Uint8Array;
  digestSRI: `sha384-${string}`;
  origin: string;
  version: string;
}

export interface StaticResource {
  uri: string;
  mediaType: string;
  bytes: Uint8Array;
  digestSRI: `sha384-${string}`;
  origin: string;
  version: string;
}

export interface CatalogBudget {
  maxResources: number;
  maxBytes: number;
}

function assertNonempty(value: string, field: string): void {
  if (value.trim().length === 0) {
    throw new CatalogError('INVALID_RESOURCE', `${field} must be nonempty.`);
  }
}

export function sha384SRI(bytes: Uint8Array): `sha384-${string}` {
  const digest = createHash('sha384').update(bytes).digest('base64');
  return `sha384-${digest}`;
}

export class StaticResourceCatalog {
  readonly #resources = new Map<string, Omit<StaticResource, 'bytes'> & { bytes: Uint8Array }>();

  constructor(inputs: readonly StaticResourceInput[]) {
    for (const input of inputs) {
      for (const [field, value] of Object.entries({
        uri: input.uri, mediaType: input.mediaType, origin: input.origin, version: input.version,
      })) assertNonempty(value, field);
      if (this.#resources.has(input.uri)) {
        throw new CatalogError('DUPLICATE_RESOURCE', `Duplicate static resource: ${input.uri}`);
      }
      const actual = sha384SRI(input.bytes);
      if (actual !== input.digestSRI) {
        throw new CatalogError(
          'INTEGRITY_MISMATCH',
          `Static resource ${input.uri} has ${actual}; expected ${input.digestSRI}.`,
        );
      }
      this.#resources.set(input.uri, { ...input, bytes: Uint8Array.from(input.bytes) });
    }
  }

  openSession(budget: CatalogBudget): CatalogSession {
    if (!Number.isSafeInteger(budget.maxResources) || budget.maxResources <= 0 ||
        !Number.isSafeInteger(budget.maxBytes) || budget.maxBytes <= 0) {
      throw new CatalogError('INVALID_RESOURCE', 'Catalog budgets must be positive safe integers.');
    }
    return new CatalogSession(this.#resources, Object.freeze({ ...budget }));
  }
}

export class CatalogSession {
  #resourcesUsed = 0;
  #bytesUsed = 0;

  constructor(
    private readonly resources: ReadonlyMap<string, Omit<StaticResource, 'bytes'> & { bytes: Uint8Array }>,
    private readonly budget: CatalogBudget,
  ) {}

  get usage(): Readonly<{ resources: number; bytes: number }> {
    return Object.freeze({ resources: this.#resourcesUsed, bytes: this.#bytesUsed });
  }

  resolve(uri: string): StaticResource {
    const entry = this.resources.get(uri);
    if (!entry) {
      throw new CatalogError('RESOURCE_NOT_FOUND', `Static resource is not installed: ${uri}`);
    }
    if (this.#resourcesUsed + 1 > this.budget.maxResources ||
        this.#bytesUsed + entry.bytes.byteLength > this.budget.maxBytes) {
      throw new CatalogError('RESOURCE_BUDGET_EXCEEDED', `Static resource budget exceeded at ${uri}.`);
    }
    this.#resourcesUsed += 1;
    this.#bytesUsed += entry.bytes.byteLength;
    return { ...entry, bytes: Uint8Array.from(entry.bytes) };
  }
}

/** Adapt one budgeted catalog session to an isolated, offline JSON-LD loader. */
export function catalogDocumentLoader(session: CatalogSession): DocumentLoader {
  return async (url: string) => {
    const resource = session.resolve(url);
    if (resource.mediaType !== 'application/json' &&
        resource.mediaType !== 'application/ld+json' &&
        !resource.mediaType.endsWith('+json')) {
      throw new CatalogError(
        'INVALID_RESOURCE',
        `JSON-LD resource ${url} has unsupported media type ${resource.mediaType}.`,
      );
    }
    let document: unknown;
    try {
      const source = new TextDecoder('utf-8', { fatal: true }).decode(resource.bytes);
      document = JSON.parse(source) as unknown;
    } catch (error) {
      throw new CatalogError(
        'INVALID_RESOURCE',
        `JSON-LD resource ${url} is not valid UTF-8 JSON: ${String(error)}.`,
      );
    }
    return { contextUrl: null, document, documentUrl: url };
  };
}

File: packages/core-py/qi_vc_core/reliance/catalog.py
# SPDX-License-Identifier: Apache-2.0
from __future__ import annotations

import base64
import hashlib
from collections.abc import Mapping
from dataclasses import dataclass
from types import MappingProxyType
from typing import Literal, Protocol

CatalogErrorCode = Literal[
    "DUPLICATE_RESOURCE",
    "INTEGRITY_MISMATCH",
    "INVALID_RESOURCE",
    "RESOURCE_NOT_FOUND",
    "RESOURCE_BUDGET_EXCEEDED",
]
_MAX_SAFE_INTEGER = 2**53 - 1


class CatalogError(ValueError):
    def __init__(self, code: CatalogErrorCode, message: str) -> None:
        super().__init__(message)
        self.code = code


@dataclass(frozen=True)
class StaticResource:
    uri: str
    media_type: str
    content: bytes
    digest_sri: str
    origin: str
    version: str


@dataclass(frozen=True)
class CatalogBudget:
    max_resources: int
    max_bytes: int


def sha384_sri(content: bytes) -> str:
    digest = hashlib.sha384(content).digest()
    return "sha384-" + base64.b64encode(digest).decode("ascii")


class StaticResourceCatalog:
    def __init__(self, inputs: list[StaticResource]) -> None:
        resources: dict[str, StaticResource] = {}
        for item in inputs:
            for field in ("uri", "media_type", "origin", "version"):
                value = getattr(item, field)
                if not value.strip():
                    raise CatalogError("INVALID_RESOURCE", f"{field} must be nonempty.")
            if item.uri in resources:
                raise CatalogError(
                    "DUPLICATE_RESOURCE", f"Duplicate static resource: {item.uri}"
                )
            content = bytes(item.content)
            actual = sha384_sri(content)
            if actual != item.digest_sri:
                raise CatalogError(
                    "INTEGRITY_MISMATCH",
                    f"Static resource {item.uri} has {actual}; "
                    f"expected {item.digest_sri}.",
                )
            resources[item.uri] = StaticResource(
                uri=item.uri,
                media_type=item.media_type,
                content=content,
                digest_sri=item.digest_sri,
                origin=item.origin,
                version=item.version,
            )
        self._resources: Mapping[str, StaticResource] = MappingProxyType(resources)

    def open_session(self, budget: CatalogBudget) -> CatalogSession:
        if (
            isinstance(budget.max_resources, bool)
            or isinstance(budget.max_bytes, bool)
            or not isinstance(budget.max_resources, int)
            or not isinstance(budget.max_bytes, int)
            or budget.max_resources <= 0
            or budget.max_bytes <= 0
            or budget.max_resources > _MAX_SAFE_INTEGER
            or budget.max_bytes > _MAX_SAFE_INTEGER
        ):
            raise CatalogError(
                "INVALID_RESOURCE",
                "Catalog budgets must be positive safe integers.",
            )
        return CatalogSession(self._resources, budget)


class CatalogSession:
    def __init__(
        self, resources: Mapping[str, StaticResource], budget: CatalogBudget
    ) -> None:
        self._resources = resources
        self._budget = budget
        self._resources_used = 0
        self._bytes_used = 0

    @property
    def usage(self) -> tuple[int, int]:
        return self._resources_used, self._bytes_used

    def resolve(self, uri: str) -> StaticResource:
        item = self._resources.get(uri)
        if item is None:
            raise CatalogError(
                "RESOURCE_NOT_FOUND", f"Static resource is not installed: {uri}"
            )
        if (
            self._resources_used + 1 > self._budget.max_resources
            or self._bytes_used + len(item.content) > self._budget.max_bytes
        ):
            raise CatalogError(
                "RESOURCE_BUDGET_EXCEEDED",
                f"Static resource budget exceeded at {uri}.",
            )
        self._resources_used += 1
        self._bytes_used += len(item.content)
        return StaticResource(**vars(item))


class ResourceResolver(Protocol):
    """Anything that resolves pinned resources by URI."""

    def resolve(self, uri: str) -> StaticResource: ...


class MemoizingResolver:
    """Resolve each distinct URI at most once, so a budget counts distinct resources."""

    def __init__(self, inner: ResourceResolver) -> None:
        self._inner = inner
        self._cache: dict[str, StaticResource | CatalogError] = {}

    @property
    def resolved(self) -> frozenset[str]:
        return frozenset(self._cache)

    def resolve(self, uri: str) -> StaticResource:
        if uri not in self._cache:
            try:
                self._cache[uri] = self._inner.resolve(uri)
            except CatalogError as error:
                self._cache[uri] = error
        entry = self._cache[uri]
        if isinstance(entry, CatalogError):
            raise entry
        return StaticResource(**vars(entry))
